### PR TITLE
Compliance query testing

### LINKED ISSUE #270: [Enhancement] New endpoint for compliance reporting
Adding in a new endpoint for getting summarized transaction data for use with upload compliance report generation

### PR DESCRIPTION
**Pull request by kauberry**
### Description

Adding in functionality for more streamlined transaction summary info for compliance reporting

### Issues Resolved

#270 
### Check List

- [ ] All tests pass. See <https://github.com/pacifica/pacifica-docs/blob/master/development/TESTING.md>
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

### CODE REVIEW

File: tests/rest/transactioninfo_test.py [hunk 1]
```diff
@@ -116,6 +116,10 @@ def test_transactioninfo_api(self):
             url='{0}/transactioninfo/multisearch?instrument_group_id={1}&project_id=1234a'.format(self.url, 1001))
         self.assertEqual(req.status_code, 200)
 
+        req = requests.get(
+            url='{0}/transactioninfo/summary/{1}/{2}'.format(self.url, '2015-07-01', '2019-07-30'))
```
Comment (dmlb2000) [inline tests/rest/transactioninfo_test.py]: @kauberry I'm wondering if your testing isn't catching the state of the metadata, the `updated` fields get updated to the current time when the Travis CI pipeline gets run. So, maybe set your end date to a calculated date that always returns tomorrow?

**Comment by kauberry** [comment]
Ah. That would definitely do it. Thanks for the insight!

Thanks,
Ken

Ken J. Auberry, <mailto:Kenneth.Auberry@pnnl.gov> Senior Data Scientist
Molecular Sciences Computing Facility
Environmental Molecular Sciences Laboratory<http://www.emsl.pnnl.gov/>
email: Kenneth.Auberry@pnnl.gov<mailto:Kenneth.Auberry@pnnl.gov>
tel: <tel:+1-509-371-6442> <tel:509.371.6442> 509<tel:509.371.6442>.<tel:509.371.6442><tel:509.371.6442>371<tel:509.371.6442><tel:509.371.6442>.<tel:509.371.6442><tel:509.371.6442>6442<tel:509.371.6442>
cell: <tel:+1-509-371-6442> <tel:509.371.6442> 509<tel:509.845.3010>.<tel:509.845.3010><tel:509.371.6442>8<tel:509.845.3010>45<tel:509.845.3010><tel:509.371.6442>.<tel:509.845.3010><tel:509.371.6442>3<tel:509.845.3010>010<tel:509.845.3010>

On Jan 26, 2020, at 12:09 PM, David Brown <notifications@github.com> wrote:

﻿

@dmlb2000 commented on this pull request.

________________________________

In tests/rest/transactioninfo_test.py<https://protect2.fireeye.com/v1/url?k=b4fe61d6-e84b5f19-b4fe4bc3-0cc47adc5e60-4cefbd1731d292b9&q=1&e=0832f71c-9363-4d84-b632-6fc281923e25&u=https%3A%2F%2Fgithub.com%2Fpacifica%2Fpacifica-metadata%2Fpull%2F271%23discussion_r371026066>:

> @@ -116,6 +116,10 @@ def test_transactioninfo_api(self):
             url='{0}/transactioninfo/multisearch?instrument_group_id={1}&project_id=1234a'.format(self.url, 1001))
         self.assertEqual(req.status_code, 200)

+        req = requests.get(
+            url='{0}/transactioninfo/summary/{1}/{2}'.format(self.url, '2015-07-01', '2019-07-30'))


@kauberry<https://protect2.fireeye.com/v1/url?k=0bba0fc6-570f3109-0bba25d3-0cc47adc5e60-dc1bef308c5b314f&q=1&e=0832f71c-9363-4d84-b632-6fc281923e25&u=https%3A%2F%2Fgithub.com%2Fkauberry> I'm wondering if your testing isn't catching the state of the metadata, the updated fields get updated to the current time when the Travis CI pipeline gets run. So, maybe set your end date to a calculated date that always returns tomorrow?

File: pacifica/metadata/rest/reporting_queries/transaction_reporting.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @staticmethod
+    def get_transaction_date_range_details(start_date, end_date):
+        """Return a transaction set grouped on instrument and project for a given time span."""
+        # pylint: disable=no-member
+        transaction_query = (
+            TransSIP().select(
+                fn.Count(TransSIP.id).alias('transaction_count'),
+                fn.Min(TransSIP.updated).alias('earliest_upload_date'),
+                fn.Max(TransSIP.updated).alias('latest_upload_date'),
+                fn.Min(TransSIP.submitter).alias('uploaded_by_id'),
```
[the remark this answers is not in the document]
Comment (kauberry) [reply pacifica/metadata/rest/reporting_queries/transaction_reporting.py]: If you're talking about that last one with the submitter, that's just the way that you pick a single value for a grouping. If you're talking about the dates, that just gets me the spread of dates within the grouping. I could group on the uploader id as well, but that just complicates things.

[the remark this answers is not in the document]
Comment (dmlb2000) [reply pacifica/metadata/rest/reporting_queries/transaction_reporting.py]: I am talking about picking the minimum of the submitters. Maybe I'm not understanding what you are using it for. Selecting counts of id, min and max of updated, project_id and instrument_id seems fine. What's the purpose of uploaded_by_id? How are you using it on your end?

[the remark this answers is not in the document]
Comment (dmlb2000) [reply pacifica/metadata/rest/reporting_queries/transaction_reporting.py]: Did you mean to add `TransSIP.submitter` to the `group_by` statement?

[the remark this answers is not in the document]
Comment (dmlb2000) [reply pacifica/metadata/rest/reporting_queries/transaction_reporting.py]: @kauberry Thanks, for the change.

**Comment by dmlb2000** [comment]
@kauberry Thanks for the really clean pull request. Just wanted to let you know, I checked with @GitBytes and I applied this change to the current production version of Metadata running on NExUS. The patch applied cleanly with no issues and I make v0.12.3.1 that includes this endpoint as well as applying it to master.

 * https://travis-ci.org/pacifica/pacifica-metadata/builds/643213899?utm_medium=notification&utm_source=github_status
 * https://travis-ci.org/pacifica/pacifica-metadata/builds/643203623?utm_medium=notification&utm_source=github_status

**Pull request merged**